roachtest: rudimentary mixed version test

tbg opened the pull request:
The bigger challenge is to find out how to achieve significant test
coverage. We should likely run at least everything `workload` has to
offer.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23205)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Interesting stuff in `workload` that you can run right now are `kv`, `tpcc` and splits. The last seems particularly interesting in a mixed cluster setting given that splits and mixed-version clusters are both rare. Would also be good to have a workload the forces replica and lease rebalance (perhaps a loop that is scattering ranges).  All of this is brainstorming for a future PR. 

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/version.go, line 50 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6RfwHCGsBeXLjxvGqY:-L6RfwHD4aA7olE5Dp6o:b-ara7yi) ([raw file](https://github.com/cockroachdb/cockroach/blob/59244d15f91360847b6784422538b1f3e1bf6148/pkg/cmd/roachtest/version.go#L50)):*
> ```Go
> 		c.Put(ctx, cockroach, "./cockroach", c.Range(nodes/2+1, nodes))
> 		// It happens in a completely different place, but the startup order makes
> 		// this cluster run at `version`. Pinky promise. (We should plumb `PGUrl`
> ```

I was imagining you'd start all of the nodes at `version`, then do a rolling restart. Also, should you be bumping the cluster version at some point? I think we want to test both modes, where we don't bump the cluster version and where we do bump the cluster version. 

---

*[pkg/cmd/roachtest/version.go, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6Rfc-kP4iVCWJmMOTe:-L6Rfc-kP4iVCWJmMOTf:b-41lg1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/59244d15f91360847b6784422538b1f3e1bf6148/pkg/cmd/roachtest/version.go#L51)):*
> ```Go
> 		// It happens in a completely different place, but the startup order makes
> 		// this cluster run at `version`. Pinky promise. (We should plumb `PGUrl`
> 		// and then connect and check).
> ```

I think you can use `cluster.Conn` for this purpose now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-:-L6RgIpS8QQFdFwHdJT-:b-llmu2u)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Pushed some more updates. I'm pretty sure this test is going to be flaky as is (in fact, looking at a flake right now) but once it isn't any more, it should be a pretty good start.

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/cmd/roachtest/version.go, line 50 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6RfwHCGsBeXLjxvGqY:-L6UlPAJ99NvHTBEowJC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59244d15f91360847b6784422538b1f3e1bf6148/pkg/cmd/roachtest/version.go#L50)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I was imagining you'd start all of the nodes at `version`, then do a rolling restart. Also, should you be bumping the cluster version at some point? I think we want to test both modes, where we don't bump the cluster version and where we do bump the cluster version. 
</blockquote></details>

Done.

---

*[pkg/cmd/roachtest/version.go, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6Rfc-kP4iVCWJmMOTe:-L6UlPYfMyO9Q5EfJAe3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59244d15f91360847b6784422538b1f3e1bf6148/pkg/cmd/roachtest/version.go#L51)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think you can use `cluster.Conn` for this purpose now.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-:-L6Us3Yv5sPYSSDL19-b:br73ak)*
<!-- Sent from Reviewable.io -->

petermattis commented:
This is looking really nice. 

---

Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/cmd/roachtest/version.go, line 92 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6b8o3hH05794o_ilLz:-L6b8o3hH05794o_ilM-:b-dxq8b1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a73f2c60843d3bd0e7c2c7ea8f6712a70fffb65/pkg/cmd/roachtest/version.go#L92)):*
> ```Go
> 				// TODO(tschottdorf): It's a bit silly that we use a dedicated load gen
> 				// machine here, but `c.Stop` calls `roachprod stop` and that kills
> 				// *everything* on that machine, not just the cockroach process.
> ```

I love that you're documenting these deficiencies.

---

*[pkg/cmd/roachtest/version.go, line 98 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6b8D9GsaegaSy3LYTG:-L6b8D9GsaegaSy3LYTH:b6tt19z) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a73f2c60843d3bd0e7c2c7ea8f6712a70fffb65/pkg/cmd/roachtest/version.go#L98)):*
> ```Go
> 
> 		m.Go(func() error {
> 			// TODO(tschottdorf): why am I not seeing the output?
> ```

Not sure. Seems like you should be seeing output here, prefixed with `upgrader:`. 

---

*[pkg/cmd/roachtest/version.go, line 137 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6b94POBbM4VCfzPTR9:-L6b94POBbM4VCfzPTRA:by32xzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a73f2c60843d3bd0e7c2c7ea8f6712a70fffb65/pkg/cmd/roachtest/version.go#L137)):*
> ```Go
> 			for i := 1; i <= nodes; i++ {
> 				l.printf("upgrading", i)
> 				c.Stop(ctx, c.Node(i))
> ```

All alternative to calling `Stop` on a node (which kills everything) is to do something like `c.Run(ctx, c.Node(i), fmt.Sprintf(`./cockroach quit --url {pgurl:%d}`, i))`. This doesn't quite work because `quit` doesn't take a `--url` flag. I think we'd need to add a `{port}` variable to `roachprod`. In addition to only stopping cockroach, another advantage to using `quit` is that it would exercise draining.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-:-L6b8whAHLJXwRqit_od:bb2573i)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Ready for another look. Mostly works locally, and mostly works remotely (though when something fails it's so hard to debug; we really need to invest heavily in that or nobody wants to ever touch this again -- I can already feel that coming on in myself).

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/cmd/roachtest/version.go, line 92 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6b8o3hH05794o_ilLz:-L6ceF3JVfbiFzGVLQOz:b-td9vo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a73f2c60843d3bd0e7c2c7ea8f6712a70fffb65/pkg/cmd/roachtest/version.go#L92)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I love that you're documenting these deficiencies.
</blockquote></details>

Thanks. To be honest, I think we need to quickly invest in the UX here. I'm already dreading fixing up this test because from now on, it's all just fighting the test infra.

1. the test is immediately flaky with 5 nodes because the first node doesn't get a join flag, and so restarting it, there's no guarantee it can connect to Gossip (I think.

---

*[pkg/cmd/roachtest/version.go, line 98 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6b8D9GsaegaSy3LYTG:-L6eClT1UB6uEXDduRuV:b-vnaskd) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a73f2c60843d3bd0e7c2c7ea8f6712a70fffb65/pkg/cmd/roachtest/version.go#L98)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Not sure. Seems like you should be seeing output here, prefixed with `upgrader:`. 
</blockquote></details>

Double checked, I was seeing it but it got drowned out by all the load gens. Made some adjustments.

---

*[pkg/cmd/roachtest/version.go, line 137 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6b94POBbM4VCfzPTR9:-L6eCox-C2NrESqeAgHS:b4e7xn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a73f2c60843d3bd0e7c2c7ea8f6712a70fffb65/pkg/cmd/roachtest/version.go#L137)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

All alternative to calling `Stop` on a node (which kills everything) is to do something like `c.Run(ctx, c.Node(i), fmt.Sprintf(`./cockroach quit --url {pgurl:%d}`, i))`. This doesn't quite work because `quit` doesn't take a `--url` flag. I think we'd need to add a `{port}` variable to `roachprod`. In addition to only stopping cockroach, another advantage to using `quit` is that it would exercise draining.  
</blockquote></details>

Done, but unfortunately `quit` isn't great in that it doesn't guarantee that you can immediately restart and not have the old process get in your way. Using a quit-kill combo for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-:-L6eCvsg7UuvE1pr32En:b2wqa7c)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/version.go, line 92 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6b8o3hH05794o_ilLz:-L6g3VcEZS4Kwqn0TK_z:b2tksyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a73f2c60843d3bd0e7c2c7ea8f6712a70fffb65/pkg/cmd/roachtest/version.go#L92)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Thanks. To be honest, I think we need to quickly invest in the UX here. I'm already dreading fixing up this test because from now on, it's all just fighting the test infra.

1. the test is immediately flaky with 5 nodes because the first node doesn't get a join flag, and so restarting it, there's no guarantee it can connect to Gossip (I think.
</blockquote></details>

As long as the first node was part of the cluster previously, restarting it should be fine as it will have written the addresses of the other nodes in the cluster to disk. Maybe.

I agree that the UX for some of these operations needs to be improved, though I'd like to see a chaos-style test first so that we have that experience in hand. I absolutely do not think we're "done" with the UX here and want to see continued investment in both the short and medium term. 

---

*[pkg/cmd/roachtest/version.go, line 150 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6g30PUCWnsNgB-uDfe:-L6g30PUCWnsNgB-uDff:b-ih3aup) ([raw file](https://github.com/cockroachdb/cockroach/blob/c72988db0af0d13d10ed7b8e8fe3f759452ace3e/pkg/cmd/roachtest/version.go#L150)):*
> ```Go
> 
> 			stop := func(node int) {
> 				port := 26257 + portMul*(node-1)
> ```

The `portMul` stuff is ugly! Please add a TODO (or file an issue) that we should add a `{port}` variable expansion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/version.go, line 92 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6b8o3hH05794o_ilLz:-L6gAr5eHVNybDL8Kyxc:bsw5f11) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a73f2c60843d3bd0e7c2c7ea8f6712a70fffb65/pkg/cmd/roachtest/version.go#L92)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

As long as the first node was part of the cluster previously, restarting it should be fine as it will have written the addresses of the other nodes in the cluster to disk. Maybe.

I agree that the UX for some of these operations needs to be improved, though I'd like to see a chaos-style test first so that we have that experience in hand. I absolutely do not think we're "done" with the UX here and want to see continued investment in both the short and medium term. 
</blockquote></details>

Idea: we could extend the `ROACHPROD` env variable to allow tagging processes with a string so that they can later be stopped. Something like:

```
roachprod ssh local:1 --tag "foo" -- sleep 1000
...
roachprod stop local:1 --tag "foo"  
```

Internally, the first command would add the env var `ROACHPROD=1/foo`. The second command would look for processes with the env var `ROACHPROD=(1)/foo`. Thoughts?

---

*[pkg/cmd/roachtest/version.go, line 150 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6g30PUCWnsNgB-uDfe:-L6gA4bOWZlCWiutn3io:btfapy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c72988db0af0d13d10ed7b8e8fe3f759452ace3e/pkg/cmd/roachtest/version.go#L150)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The `portMul` stuff is ugly! Please add a TODO (or file an issue) that we should add a `{port}` variable expansion.
</blockquote></details>

https://github.com/cockroachdb/roachprod/pull/118

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/version.go, line 92 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6b8o3hH05794o_ilLz:-L6tvnPVoQ454XeVuT65:b-socgoy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a73f2c60843d3bd0e7c2c7ea8f6712a70fffb65/pkg/cmd/roachtest/version.go#L92)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Idea: we could extend the `ROACHPROD` env variable to allow tagging processes with a string so that they can later be stopped. Something like:

```
roachprod ssh local:1 --tag "foo" -- sleep 1000
...
roachprod stop local:1 --tag "foo"  
```

Internally, the first command would add the env var `ROACHPROD=1/foo`. The second command would look for processes with the env var `ROACHPROD=(1)/foo`. Thoughts?
</blockquote></details>

Sorry, that first comment was a work in progress until I forgot about it. The join flags here are fine but I continue to be worried that these tests as written are already really unpleasant to debug since the log output is mostly a jumbled mess and it takes  ~100 steps to get from the test to where the error is occurring, and often you have to manually insert a string of printfs since the errors don't usually come annotated with a location in the source. (I know we agree on these things, just wanted to write the comment I wanted to have left above).

The idea to extend the env var SGTM. We could also just extract the PID, but with an eye for long-running tests (for which the coordinator does not want to hang on to a session) the env var seems better. But it has sort of the same problem that it assumes that it owns the tag. For example, you won't be able to run two instances of the same test side by side locally. I would say that's a non-issue, though I did it just the other day...

---

*[pkg/cmd/roachtest/version.go, line 150 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205#-L6g30PUCWnsNgB-uDfe:-L6twdKl7sG4KrgORikY:b28xasy) ([raw file](https://github.com/cockroachdb/cockroach/blob/c72988db0af0d13d10ed7b8e8fe3f759452ace3e/pkg/cmd/roachtest/version.go#L150)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

https://github.com/cockroachdb/roachprod/pull/118
</blockquote></details>

Done. It will be broken for everyone who doesn't upgrade roachprod, though I guess that's just what it is for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23205)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/cmd/roachtest/version.go, line 92 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23205#-L6b8o3hH05794o_ilLz:-L6wbJjL32cqB_Fq4EWD:bxl6ipq) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a73f2c60843d3bd0e7c2c7ea8f6712a70fffb65/pkg/cmd/roachtest/version.go#L92)):*
> The join flags here are fine but I continue to be worried that these tests as written are already really unpleasant to debug since the log output is mostly a jumbled mess and it takes ~100 steps to get from the test to where the error is occurring, and often you have to manually insert a string of printfs since the errors don't usually come annotated with a location in the source. (I know we agree on these things, just wanted to write the comment I wanted to have left above).

As you know, suggestions are welcome. So are the criticisms and critiques, but suggestions are more actionable.

Running two instances of the same test side-by-side on the same cluster seems difficult, though I think the `--tag` option can handle that if we preface the tag with the test name. 

Let me whip up support for the `--tag` option in `roachprod`. Shouldn't be difficult to do.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23205)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/cmd/roachtest/version.go, line 150 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23205#-L6g30PUCWnsNgB-uDfe:-L6wcBNm8en9DlM3609w:bpijwnk) ([raw file](https://github.com/cockroachdb/cockroach/blob/c72988db0af0d13d10ed7b8e8fe3f759452ace3e/pkg/cmd/roachtest/version.go#L150)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Done. It will be broken for everyone who doesn't upgrade roachprod, though I guess that's just what it is for now.
</blockquote></details>

We kind of expect frequent `roachprod` upgrades right now.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23205)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/version.go, line 92 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23205#-L6b8o3hH05794o_ilLz:-L6zE3ex4r24kvGy0nHA:boq0jw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a73f2c60843d3bd0e7c2c7ea8f6712a70fffb65/pkg/cmd/roachtest/version.go#L92)):*
> if we preface the tag with the test name.

I was actually running the *same* test side by side. But as I said, I think I can just not do that again and live.

> As you know, suggestions are welcome. So are the criticisms and critiques, but suggestions are more actionable.

I think we just need to TLC this. A good first step is making all logging glog-like (i.e. with file:line numbers) and using `pkg/errors` throughout so that errors have the file:line of their creation (and then there are some errors that are hideously opaque an difficult to improve at the source, so we should just wrap them with a hint, the darwin-binary-to-linux-machine one for example). I guess we can just use our logging package in `roachprod` and `roachtest` alike (or `glog` in roachprod`). Second iterate on what goes on the screen. For example, the `./cockroach start` invocations aren't printed. That strikes me as unfortunate but is surely easy to fix. And then there's progress tracking without cluttering the test which I suppose we track [here](https://github.com/cockroachdb/cockroach/issues/23336).

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23205)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/cmd/roachtest/version.go, line 92 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23205#-L6b8o3hH05794o_ilLz:-L7-Z7AO7N-Dk9MummLw:btyoctz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a73f2c60843d3bd0e7c2c7ea8f6712a70fffb65/pkg/cmd/roachtest/version.go#L92)):*
> A good first step is making all logging glog-like (i.e. with file:line numbers)

The standard `log` package can be configured to display `file:line`. Let's file a bug about this and continue the discussion there as I'm not quite sure what messages you want `file:line` info for.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23205)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/cmd/roachtest/version.go, line 92 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23205#-L6b8o3hH05794o_ilLz:-L73L--I8c62JbrimZo9:b-qdgsj4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a73f2c60843d3bd0e7c2c7ea8f6712a70fffb65/pkg/cmd/roachtest/version.go#L92)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

> A good first step is making all logging glog-like (i.e. with file:line numbers)

The standard `log` package can be configured to display `file:line`. Let's file a bug about this and continue the discussion there as I'm not quite sure what messages you want `file:line` info for.
</blockquote></details>

Let me hold off on that, maybe your recent improvements have made that unnecessary. I was interested in this mostly in the context of tracking down the origin of errors or log messages.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23205)*
<!-- Sent from Reviewable.io -->

tbg committed:
roachtest: rudimentary mixed version test

Upgrade from 1.1.x into the current binary while running TPCC and KV
with splits.

As a nice side effect, running this over longer periods of time
should be an easy way of gauging the performance improvements
of 2.0 and the impact of rolling restarts on performance (though
workload likely isn't smart enough just yet to avoid trying down nodes
repeatedly).

Release note: None
